crates/compiler/src/error_listener.rs

```diff
@@ -52,3 +108,127 @@ pub enum DiagnosticSeverity {
     /// to fix, but are unlikely to cause problems.
     Info,
 }
+
+pub(crate) struct LexerErrorListener {
+    pub(crate) diagnostics: RefCell<Vec<Diagnostic>>,
+    file_name: String,
+}
+
+impl LexerErrorListener {
+    pub(crate) fn new(file_name: String) -> Self {
+        Self {
+            file_name,
+            diagnostics: Default::default(),
+        }
+    }
+}
+
+impl<'input, T: Recognizer<'input>> ErrorListener<'input, T> for LexerErrorListener {
+    fn syntax_error(
+        &self,
+        _recognizer: &T,
+        _offending_symbol: Option<&<T::TF as TokenFactory<'input>>::Inner>,
+        line: isize,
+        column: isize,
+        msg: &str,
+        _error: Option<&ANTLRError>,
+    ) {
+        let line = (line - 1) as usize;
+        let column = column as usize;
+        let range = Position {
+            line,
+            character: column,
+        }..=Position {
+            line,
+            character: column + 1,
+        };
+        self.diagnostics.borrow_mut().push(
+            Diagnostic::from_message(msg)
+                .with_range(range)
+                .with_file_name(&self.file_name),
+        );
+    }
+}
+
+pub(crate) struct ParserErrorListener {
+    pub(crate) diagnostics: Rc<RefCell<Vec<Diagnostic>>>,
+    /// Not in original implementation, but needed because
+    /// [`Token::get_source`] is not implemented by antlr4rust.
+    /// So we take the entire file instead of just the file name
+    /// and extract the lines ourselves.
+    file: File,
+}
+
+impl ParserErrorListener {
+    pub fn new(file: File) -> Self {
+        Self {
+            diagnostics: Default::default(),
+            file,
+        }
+    }
+}
+
+impl<'input, T: Recognizer<'input>> ErrorListener<'input, T> for ParserErrorListener {
+    fn syntax_error(
+        &self,
+        _recognizer: &T,
+        offending_symbol: Option<&<T::TF as TokenFactory<'input>>::Inner>,
+        line: isize,
+        column: isize,
+        msg: &str,
+        _error: Option<&ANTLRError>,
+    ) {
+        let range = Position {
+            line: (line - 1) as usize,
```

Comment:
this -1 +1 pattern is repeating itself, what does it mean?

Reply:
For +1 see my last comment, for -1 I can only speculate as much as you

Reply:
Worth noting is that `line - 1` is "actually" `line - 2` because we usually convert to 1 based lines for user output. That, or this is never shown to the user and thus a real `line - 1`. Idk.